### PR TITLE
[nexus] Break up db model into multiple files

### PR DESCRIPTION
**Pull request by smklein**
- moved `nexus/src/db/model.rs` into `nexus/src/db/model/mod.rs`
- migrated all DB structures into separate files

### CODE REVIEW

File: nexus/src/db/model.rs [hunk 1]
```diff
@@ -1,2866 +0,0 @@
...
-use std::convert::TryFrom;
-use std::io::Write;
-use std::net::IpAddr;
-use std::net::Ipv4Addr;
-use std::net::Ipv6Addr;
-use std::net::SocketAddr;
-use std::net::SocketAddrV6;
-use uuid::Uuid;
-
-// TODO: Break up types into multiple files
```
Comment (smklein) [inline nexus/src/db/model.rs:46]: On the bright side, this is done now

**Review by david-crespo — APPROVED** [review]
long overdue, big improvement. let's do the other ones

**Comment by smklein** [comment]
Sure, that would be okay with me. I think there are a number of ways to resolve the "grouping":

- Use a common prefix. `instance.rs` is right next to `instance_state.rs`. `vpc.rs` is right next to `vpc_route.rs`, `vpc_firewall_rule.rs`, etc.
- Have related models in the same file, as you suggested. I actually think something as broad as `networking.rs` will run into the same problem as `model.rs` did - it's still going to grow to be thousands of lines long. But, as an example, I did leave structures like `VpcSubnet` and `VpcSubnetUpdate` in the same file, because they're very tightly coupled.

**Comment by david-crespo** [comment]
I think I like the prefix idea better than the bigger files.

**Comment by davepacheco** [comment]
I prefer bigger files because they can be structured with a logical narrative instead of strictly alphabetical.  When I'm coming up to speed on something I _like_ to pull up a big file, read it top to bottom (ideally starting with a big subsystem block comment), and feel like I understand the structure of the thing.

**Comment by bnaecker** [comment]
Another option is additional layers of hierarchy in the module structure. For example, `db::model::networking::{vpc,vpc_subnet,firewall_rules,...}.rs` and similar.

**Comment by smklein** [comment]
@bnaecker I like that idea - it's a nice compromise, because it lets those of us who like small files keep them, while it also provides a natural grouping (by directory) in which we could include, for example, area-specific documentation

**Comment by davepacheco** [comment]
I don't know where to discuss the other ones in flight before we go ahead and do them...

Regarding nexus.rs: is it worth rethinking this one?  The original idea was to have a sort of business logic layer between the datastore and the HTTP endpoints that could be shared with interfaces besides than the HTTP endpoints.  The best case _for_ something like this is the logic around `instance_set_runtime()`, which is non-trivial, unrelated to the HTTP API itself, and might be used elsewhere like sagas or Nexus background tasks.

At the same time it feels like there's a ton of boilerplate here where this layer isn't adding anything.  For the various CRUD functions that are all just (1) use LookupPath to look up a resource and then (2) call a datastore function to do something with that resource, I wonder if we're better off doing that stuff directly from the HTTP entrypoints.

The datastore layer exposes a pretty low-level interface atop the database that attempts to ensure database-level integrity, but doesn't know about higher-level application-level concerns.  Examples: the idea of an `Actor` or `ConsoleSessionWithSiloId` is specific to the authn subsystem.  Maybe the way to break up Nexus is to have more of these subsystems, maybe for different groups of resources, like an "instance" submodule that would contain the equivalent of `instance_set_runtime()`, etc.  This might be clearer than a monolithic layer like nexus.rs currently provides.  This is also different from just splitting the existing Nexus functions into different files.  There's still a question of what to do for things that really have no specialized logic around them that are just basic CRUD.

**Comment by david-crespo** [comment]
I'm ambivalent about the `nexus.rs` CRUD-only functions. When I see a function that does one thing and is called once, I usually would like to see it inlined at the call site (the endpoint handler). On the other hand, I can see a maintainability argument for keeping the one-liner functions in the service layer (what I call `nexus.rs` in my head), because it means when you want to expand that logic or share it between endpoints, you don't have to complicate the diff by also doing the extraction to a function in `nexus.rs`. That's definitely the Railsy approach, where one-liner functions are the norm, in the name of having things in the "right" place. Like I said, I'm ambivalent — sometimes I feel the Rails style is overly defensive, putting too much junk in my way to guard against a hypothetical future "bad" developer that doesn't exist.

**Pull request merged**